File: aop/aop-annotation-processor/src/main/java/ru/tinkoff/kora/aop/annotation/processor/KoraAspect.java
package ru.tinkoff.kora.aop.annotation.processor;

import com.squareup.javapoet.AnnotationSpec;
import com.squareup.javapoet.CodeBlock;

import javax.lang.model.element.ExecutableElement;
import javax.lang.model.type.TypeMirror;
import java.util.List;
import java.util.Set;

public interface KoraAspect {
    Set<String> getSupportedAnnotationTypes();

    interface FieldFactory {

        String constructorParam(TypeMirror type, List<AnnotationSpec> annotations);

        String constructorInitialized(TypeMirror type, CodeBlock initializer);
    }

    sealed interface ApplyResult {
        enum Noop implements ApplyResult {INSTANCE}

        record MethodBody(CodeBlock codeBlock) implements ApplyResult {}
    }
    record AspectContext(FieldFactory fieldFactory){}

    ApplyResult apply(ExecutableElement executableElement, String superCall, AspectContext aspectContext);
}

File: aop/aop-symbol-processor/src/main/kotlin/ru/tinkoff/kora/aop/symbol/processor/KoraAspect.kt
package ru.tinkoff.kora.aop.symbol.processor

import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.AnnotationSpec
import com.squareup.kotlinpoet.CodeBlock

interface KoraAspect {
    fun getSupportedAnnotationTypes(): Set<String>
    interface FieldFactory {
        fun constructorParam(type: KSType, annotations: List<AnnotationSpec>): String
        fun constructorInitialized(type: KSType, initializer: CodeBlock): String
    }

    interface ApplyResult {
        enum class Noop : ApplyResult {
            INSTANCE
        }

        data class MethodBody(val codeBlock: CodeBlock) : ApplyResult
    }

    data class AspectContext(val fieldFactory: FieldFactory)

    fun apply(ksFunction: KSFunctionDeclaration, superCall: String, aspectContext: AspectContext): ApplyResult
}
